File: AppCheckCore/Tests/Unit/DeviceCheckProvider/GACDeviceCheckProviderTests.m
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import <XCTest/XCTest.h>

#import <OCMock/OCMock.h>
#import "FBLPromise+Testing.h"

#import "AppCheckCore/Sources/Core/Errors/GACAppCheckErrorUtil.h"
#import "AppCheckCore/Sources/DeviceCheckProvider/API/GACDeviceCheckAPIService.h"
#import "AppCheckCore/Sources/DeviceCheckProvider/GACDeviceCheckTokenGenerator.h"
#import "AppCheckCore/Sources/Public/AppCheckCore/GACAppCheckToken.h"
#import "AppCheckCore/Sources/Public/AppCheckCore/GACDeviceCheckProvider.h"

#import "AppCheckCore/Tests/Utils/AppCheckBackoffWrapperFake/GACAppCheckBackoffWrapperFake.h"

GAC_DEVICE_CHECK_PROVIDER_AVAILABILITY
@interface GACDeviceCheckProvider (Tests)

- (instancetype)initWithAPIService:(id<GACDeviceCheckAPIServiceProtocol>)APIService
              deviceTokenGenerator:(id<GACDeviceCheckTokenGenerator>)deviceTokenGenerator
                    backoffWrapper:(id<GACAppCheckBackoffWrapperProtocol>)backoffWrapper;

@end

GAC_DEVICE_CHECK_PROVIDER_AVAILABILITY
@interface GACDeviceCheckProviderTests : XCTestCase

@property(nonatomic) GACDeviceCheckProvider *provider;
@property(nonatomic) id fakeAPIService;
@property(nonatomic) id fakeTokenGenerator;
@property(nonatomic) GACAppCheckBackoffWrapperFake *fakeBackoffWrapper;

@end

@implementation GACDeviceCheckProviderTests

- (void)setUp {
  [super setUp];

  self.fakeAPIService = OCMProtocolMock(@protocol(GACDeviceCheckAPIServiceProtocol));
  self.fakeTokenGenerator = OCMProtocolMock(@protocol(GACDeviceCheckTokenGenerator));

  self.fakeBackoffWrapper = [[GACAppCheckBackoffWrapperFake alloc] init];
  // Don't backoff by default.
  self.fakeBackoffWrapper.isNextOperationAllowed = YES;

  self.provider = [[GACDeviceCheckProvider alloc] initWithAPIService:self.fakeAPIService
                                                deviceTokenGenerator:self.fakeTokenGenerator
                                                      backoffWrapper:self.fakeBackoffWrapper];
}

- (void)tearDown {
  self.provider = nil;
  self.fakeAPIService = nil;
  self.fakeTokenGenerator = nil;
  self.fakeBackoffWrapper = nil;
}

- (void)testGetTokenSuccess {
  // 1. Expect GACDeviceCheckTokenGenerator.isSupported.
  OCMExpect([self.fakeTokenGenerator isSupported]).andReturn(YES);

  // 2. Expect device token to be generated.
  NSData *deviceToken = [NSData data];
  id generateTokenArg = [OCMArg invokeBlockWithArgs:deviceToken, [NSNull null], nil];
  OCMExpect([self.fakeTokenGenerator generateTokenWithCompletionHandler:generateTokenArg]);

  // 3. Expect FAA token to be requested.
  GACAppCheckToken *validToken = [[GACAppCheckToken alloc] initWithToken:@"valid_token"
                                                          expirationDate:[NSDate distantFuture]
                                                          receivedAtDate:[NSDate date]];
  OCMExpect([self.fakeAPIService appCheckTokenWithDeviceToken:deviceToken limitedUse:NO])
      .andReturn([FBLPromise resolvedWith:validToken]);
  OCMReject([self.fakeAPIService appCheckTokenWithDeviceToken:OCMOCK_ANY limitedUse:YES]);

  // 4. Expect backoff wrapper to be used.
  self.fakeBackoffWrapper.backoffExpectation = [self expectationWithDescription:@"Backoff"];

  // 5. Call getToken and validate the result.
  XCTestExpectation *completionExpectation =
      [self expectationWithDescription:@"completionExpectation"];
  [self.provider
      getTokenWithCompletion:^(GACAppCheckToken *_Nullable token, NSError *_Nullable error) {
        [completionExpectation fulfill];
        XCTAssertEqualObjects(token.token, validToken.token);
        XCTAssertEqualObjects(token.expirationDate, validToken.expirationDate);
        XCTAssertEqualObjects(token.receivedAtDate, validToken.receivedAtDate);
        XCTAssertNil(error);
      }];

  [self waitForExpectations:@[ self.fakeBackoffWrapper.backoffExpectation, completionExpectation ]
                    timeout:0.5
               enforceOrder:YES];

  // 6. Verify.
  XCTAssertNil(self.fakeBackoffWrapper.operationError);
  GACAppCheckToken *wrapperResult =
      [self.fakeBackoffWrapper.operationResult isKindOfClass:[GACAppCheckToken class]]
          ? self.fakeBackoffWrapper.operationResult
          : nil;
  XCTAssertEqualObjects(wrapperResult.token, validToken.token);

  OCMVerifyAll(self.fakeAPIService);
  OCMVerifyAll(self.fakeTokenGenerator);
}

- (void)testGetTokenWhenDeviceCheckIsNotSupported {
  NSError *expectedError =
      [GACAppCheckErrorUtil unsupportedAttestationProvider:@"DeviceCheckProvider"];

  // 0.1. Expect backoff wrapper to be used.
  self.fakeBackoffWrapper.backoffExpectation = [self expectationWithDescription:@"Backoff"];

  // 0.2. Expect default error handler to be used.
  XCTestExpectation *errorHandlerExpectation = [self expectationWithDescription:@"Error handler"];
  self.fakeBackoffWrapper.defaultErrorHandler = ^GACAppCheckBackoffType(NSError *_Nonnull error) {
    XCTAssertEqualObjects(error, expectedError);
    [errorHandlerExpectation fulfill];
    return GACAppCheckBackoffType1Day;
  };

  // 1. Expect GACDeviceCheckTokenGenerator.isSupported.
  OCMExpect([self.fakeTokenGenerator isSupported]).andReturn(NO);

  // 2. Don't expect DeviceCheck token to be generated or FAA token to be requested.
  OCMReject([self.fakeTokenGenerator generateTokenWithCompletionHandler:OCMOCK_ANY]);
  OCMReject([self.fakeAPIService appCheckTokenWithDeviceToken:OCMOCK_ANY limitedUse:NO])
      .ignoringNonObjectArgs();

  // 3. Call getToken and validate the result.
  XCTestExpectation *completionExpectation =
      [self expectationWithDescription:@"completionExpectation"];
  [self.provider
      getTokenWithCompletion:^(GACAppCheckToken *_Nullable token, NSError *_Nullable error) {
        [completionExpectation fulfill];
        XCTAssertNil(token);
        XCTAssertEqualObjects(error, expectedError);
      }];

  [self waitForExpectations:@[
    self.fakeBackoffWrapper.backoffExpectation, errorHandlerExpectation, completionExpectation
  ]
                    timeout:0.5
               enforceOrder:YES];

  // 4. Verify.
  OCMVerifyAll(self.fakeAPIService);
  OCMVerifyAll(self.fakeTokenGenerator);

  XCTAssertEqualObjects(self.fakeBackoffWrapper.operationError, expectedError);
  XCTAssertNil(self.fakeBackoffWrapper.operationResult);
}

- (void)testGetTokenWhenDeviceTokenFails {
  NSError *deviceTokenError = [NSError errorWithDomain:@"GACDeviceCheckProviderTests"
                                                  code:-1
                                              userInfo:nil];

  // 0.1. Expect backoff wrapper to be used.
  self.fakeBackoffWrapper.backoffExpectation = [self expectationWithDescription:@"Backoff"];

  // 0.2. Expect default error handler to be used.
  XCTestExpectation *errorHandlerExpectation = [self expectationWithDescription:@"Error handler"];
  self.fakeBackoffWrapper.defaultErrorHandler = ^GACAppCheckBackoffType(NSError *_Nonnull error) {
    XCTAssertEqualObjects(error, deviceTokenError);
    [errorHandlerExpectation fulfill];
    return GACAppCheckBackoffType1Day;
  };

  // 1. Expect GACDeviceCheckTokenGenerator.isSupported.
  OCMExpect([self.fakeTokenGenerator isSupported]).andReturn(YES);

  // 2. Expect device token to be generated.
  id generateTokenArg = [OCMArg invokeBlockWithArgs:[NSNull null], deviceTokenError, nil];
  OCMExpect([self.fakeTokenGenerator generateTokenWithCompletionHandler:generateTokenArg]);

  // 3. Don't expect FAA token to be requested.
  OCMReject([self.fakeAPIService appCheckTokenWithDeviceToken:OCMOCK_ANY limitedUse:NO])
      .ignoringNonObjectArgs();

  // 4. Call getToken and validate the result.
  XCTestExpectation *completionExpectation =
      [self expectationWithDescription:@"completionExpectation"];
  [self.provider
      getTokenWithCompletion:^(GACAppCheckToken *_Nullable token, NSError *_Nullable error) {
        [completionExpectation fulfill];
        XCTAssertNil(token);
        XCTAssertEqualObjects(error, deviceTokenError);
      }];

  [self waitForExpectations:@[
    self.fakeBackoffWrapper.backoffExpectation, errorHandlerExpectation, completionExpectation
  ]
                    timeout:0.5
               enforceOrder:YES];

  // 5. Verify.
  OCMVerifyAll(self.fakeAPIService);
  OCMVerifyAll(self.fakeTokenGenerator);

  XCTAssertEqualObjects(self.fakeBackoffWrapper.operationError, deviceTokenError);
  XCTAssertNil(self.fakeBackoffWrapper.operationResult);
}

- (void)testGetTokenWhenAPIServiceFails {
  NSError *APIServiceError = [NSError errorWithDomain:@"GACDeviceCheckProviderTests"
                                                 code:-1
                                             userInfo:nil];

  // 0.1. Expect backoff wrapper to be used.
  self.fakeBackoffWrapper.backoffExpectation = [self expectationWithDescription:@"Backoff"];

  // 0.2. Expect default error handler to be used.
  XCTestExpectation *errorHandlerExpectation = [self expectationWithDescription:@"Error handler"];
  self.fakeBackoffWrapper.defaultErrorHandler = ^GACAppCheckBackoffType(NSError *_Nonnull error) {
    XCTAssertEqualObjects(error, APIServiceError);
    [errorHandlerExpectation fulfill];
    return GACAppCheckBackoffType1Day;
  };

  // 1. Expect GACDeviceCheckTokenGenerator.isSupported.
  OCMExpect([self.fakeTokenGenerator isSupported]).andReturn(YES);

  // 2. Expect device token to be generated.
  NSData *deviceToken = [NSData data];
  id generateTokenArg = [OCMArg invokeBlockWithArgs:deviceToken, [NSNull null], nil];
  OCMExpect([self.fakeTokenGenerator generateTokenWithCompletionHandler:generateTokenArg]);

  // 3. Expect FAA token to be requested.
  FBLPromise *rejectedPromise = [FBLPromise pendingPromise];
  [rejectedPromise reject:APIServiceError];
  OCMExpect([self.fakeAPIService appCheckTokenWithDeviceToken:deviceToken limitedUse:NO])
      .andReturn(rejectedPromise);
  OCMReject([self.fakeAPIService appCheckTokenWithDeviceToken:OCMOCK_ANY limitedUse:YES]);

  // 4. Call getToken and validate the result.
  XCTestExpectation *completionExpectation =
      [self expectationWithDescription:@"completionExpectation"];
  [self.provider
      getTokenWithCompletion:^(GACAppCheckToken *_Nullable token, NSError *_Nullable error) {
        [completionExpectation fulfill];
        XCTAssertNil(token);
        XCTAssertEqualObjects(error, APIServiceError);
      }];

  [self waitForExpectations:@[
    self.fakeBackoffWrapper.backoffExpectation, errorHandlerExpectation, completionExpectation
  ]
                    timeout:0.5
               enforceOrder:YES];

  // 5. Verify.
  OCMVerifyAll(self.fakeAPIService);
  OCMVerifyAll(self.fakeTokenGenerator);

  XCTAssertEqualObjects(self.fakeBackoffWrapper.operationError, APIServiceError);
  XCTAssertNil(self.fakeBackoffWrapper.operationResult);
}

- (void)testGetLimitedUseTokenSuccess {
  // 1. Expect GACDeviceCheckTokenGenerator.isSupported.
  OCMExpect([self.fakeTokenGenerator isSupported]).andReturn(YES);

  // 2. Expect device token to be generated.
  NSData *deviceToken = [NSData data];
  id generateTokenArg = [OCMArg invokeBlockWithArgs:deviceToken, [NSNull null], nil];
  OCMExpect([self.fakeTokenGenerator generateTokenWithCompletionHandler:generateTokenArg]);

  // 3. Expect FAA token to be requested.
  GACAppCheckToken *validToken = [[GACAppCheckToken alloc] initWithToken:@"valid_token"
                                                          expirationDate:[NSDate distantFuture]
                                                          receivedAtDate:[NSDate date]];
  OCMExpect([self.fakeAPIService appCheckTokenWithDeviceToken:deviceToken limitedUse:YES])
      .andReturn([FBLPromise resolvedWith:validToken]);
  OCMReject([self.fakeAPIService appCheckTokenWithDeviceToken:OCMOCK_ANY limitedUse:NO]);

  // 4. Expect backoff wrapper to be used.
  self.fakeBackoffWrapper.backoffExpectation = [self expectationWithDescription:@"Backoff"];

  // 5. Call getToken and validate the result.
  XCTestExpectation *completionExpectation =
      [self expectationWithDescription:@"completionExpectation"];
  [self.provider getLimitedUseTokenWithCompletion:^(GACAppCheckToken *_Nullable token,
                                                    NSError *_Nullable error) {
    [completionExpectation fulfill];
    XCTAssertEqualObjects(token.token, validToken.token);
    XCTAssertEqualObjects(token.expirationDate, validToken.expirationDate);
    XCTAssertEqualObjects(token.receivedAtDate, validToken.receivedAtDate);
    XCTAssertNil(error);
  }];

  [self waitForExpectations:@[ self.fakeBackoffWrapper.backoffExpectation, completionExpectation ]
                    timeout:0.5
               enforceOrder:YES];

  // 6. Verify.
  XCTAssertNil(self.fakeBackoffWrapper.operationError);
  GACAppCheckToken *wrapperResult =
      [self.fakeBackoffWrapper.operationResult isKindOfClass:[GACAppCheckToken class]]
          ? self.fakeBackoffWrapper.operationResult
          : nil;
  XCTAssertEqualObjects(wrapperResult.token, validToken.token);

  OCMVerifyAll(self.fakeAPIService);
  OCMVerifyAll(self.fakeTokenGenerator);
}

#pragma mark - Backoff tests

- (void)testGetTokenBackoff {
  // 1. Configure backoff.
  self.fakeBackoffWrapper.isNextOperationAllowed = NO;
  self.fakeBackoffWrapper.backoffExpectation = [self expectationWithDescription:@"Backoff"];

  // 2. Don't expect any operations.
  OCMReject([self.fakeAPIService appCheckTokenWithDeviceToken:OCMOCK_ANY limitedUse:NO])
      .ignoringNonObjectArgs();
  OCMReject([self.fakeTokenGenerator generateTokenWithCompletionHandler:OCMOCK_ANY]);

  // 3. Call getToken and validate the result.
  XCTestExpectation *completionExpectation =
      [self expectationWithDescription:@"completionExpectation"];
  [self.provider
      getTokenWithCompletion:^(GACAppCheckToken *_Nullable token, NSError *_Nullable error) {
        [completionExpectation fulfill];
        XCTAssertNil(token);
        XCTAssertEqualObjects(error, self.fakeBackoffWrapper.backoffError);
      }];

  [self waitForExpectations:@[ self.fakeBackoffWrapper.backoffExpectation, completionExpectation ]
                    timeout:0.5
               enforceOrder:YES];

  // 4. Verify.
  OCMVerifyAll(self.fakeAPIService);
  OCMVerifyAll(self.fakeTokenGenerator);
}

@end

File: AppCheckCore/Tests/Unit/DeviceCheckProvider/GACDeviceCheckProviderTests.swift
import XCTest
#if canImport(FBLPromises)
import FBLPromises
#endif
@testable import AppCheckCore

class GACDeviceCheckAPIServiceFake: NSObject, GACDeviceCheckAPIServiceProtocol {
    var passedDeviceToken: Data?
    var passedLimitedUse: Bool?
    var appCheckTokenResult: Result<GACAppCheckToken, Error>?
    
    func appCheckToken(deviceToken: Data, limitedUse: Bool) async throws -> GACAppCheckToken {
        passedDeviceToken = deviceToken
        passedLimitedUse = limitedUse
        guard let result = appCheckTokenResult else {
            throw NSError(domain: "FakeError", code: -1, userInfo: nil)
        }
        switch result {
        case .success(let token): return token
        case .failure(let error): throw error
        }
    }
}

class GACDeviceCheckTokenGeneratorFake: NSObject, GACDeviceCheckTokenGenerator {
    var supported: Bool = true
    var generateTokenCalled = false
    var tokenToReturn: Data?
    var errorToReturn: Error?
    
    var isSupported: Bool { supported }
    
    func generateToken(completionHandler: @escaping (Data?, Error?) -> Void) {
        generateTokenCalled = true
        if let error = errorToReturn {
            completion(nil, error)
        } else if let token = tokenToReturn {
            completion(token, nil)
        } else {
            completion(nil, NSError(domain: "FakeError", code: -1, userInfo: nil))
        }
    }
}

class GACAppCheckBackoffWrapperFake: NSObject, _GACAppCheckBackoffWrapperProtocol {
    var isNextOperationAllowed: Bool = true
    var backoffError: Error = NSError(domain: "BackoffError", code: -1, userInfo: nil)
    
    var backoffExpectation: XCTestExpectation?
    var defaultErrorHandlerCalled = false
    var defaultErrorHandler: ((Error) -> GACAppCheckBackoffType)?
    
    var operationResult: Any?
    var operationError: Error?
    
    func applyBackoff(toOperation operationProvider: @escaping () -> FBLPromise<AnyObject>, errorHandler: @escaping (Error) -> GACAppCheckBackoffType) -> FBLPromise<AnyObject> {
        backoffExpectation?.fulfill()
        if isNextOperationAllowed {
            let promise = operationProvider()
            return promise.then { value in
                self.operationResult = value
                return value
            }.catch { error in
                self.operationError = error
                _ = errorHandler(error)
            }
        } else {
            return FBLPromise<AnyObject>(backoffError)
        }
    }
    
    func defaultAppCheckProviderErrorHandler() -> (Error) -> GACAppCheckBackoffType {
        return { error in
            self.defaultErrorHandlerCalled = true
            if let handler = self.defaultErrorHandler {
                return handler(error)
            }
            return .type1Day
        }
    }
}

class GACDeviceCheckProviderTests: XCTestCase {
    var provider: GACDeviceCheckProvider!
    var fakeAPIService: GACDeviceCheckAPIServiceFake!
    var fakeTokenGenerator: GACDeviceCheckTokenGeneratorFake!
    var fakeBackoffWrapper: GACAppCheckBackoffWrapperFake!
    
    override func setUp() {
        super.setUp()
        fakeAPIService = GACDeviceCheckAPIServiceFake()
        fakeTokenGenerator = GACDeviceCheckTokenGeneratorFake()
        fakeBackoffWrapper = GACAppCheckBackoffWrapperFake()
        fakeBackoffWrapper.isNextOperationAllowed = true
        
        provider = GACDeviceCheckProvider(apiService: fakeAPIService,
                                          deviceTokenGenerator: fakeTokenGenerator,
                                          backoffWrapper: fakeBackoffWrapper)
    }
    
    override func tearDown() {
        provider = nil
        fakeAPIService = nil
        fakeTokenGenerator = nil
        fakeBackoffWrapper = nil
        super.tearDown()
    }
    
    func testGetTokenSuccess() async throws {
        fakeTokenGenerator.supported = true
        let deviceToken = Data()
        fakeTokenGenerator.tokenToReturn = deviceToken
        
        let validToken = GACAppCheckToken(token: "valid_token", expirationDate: Date.distantFuture, receivedAt: Date())
        fakeAPIService.appCheckTokenResult = .success(validToken)
        
        fakeBackoffWrapper.backoffExpectation = expectation(description: "Backoff")
        
        let token = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<GACAppCheckToken, Error>) in
            provider.getToken { token, error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else if let token = token {
                    continuation.resume(returning: token)
                } else {
                    continuation.resume(throwing: NSError(domain: "TestError", code: -1, userInfo: nil))
                }
            }
        }
        
        await fulfillment(of: [fakeBackoffWrapper.backoffExpectation!], timeout: 0.5)
        
        XCTAssertEqual(token.token, validToken.token)
        XCTAssertEqual(token.expirationDate, validToken.expirationDate)
        XCTAssertEqual(token.receivedAtDate, validToken.receivedAtDate)
        
        XCTAssertNil(fakeBackoffWrapper.operationError)
        let wrapperResult = fakeBackoffWrapper.operationResult as? GACAppCheckToken
        XCTAssertEqual(wrapperResult?.token, validToken.token)
        
        XCTAssertEqual(fakeAPIService.passedDeviceToken, deviceToken)
        XCTAssertEqual(fakeAPIService.passedLimitedUse, false)
        XCTAssertTrue(fakeTokenGenerator.generateTokenCalled)
    }
    
    func testGetTokenWhenDeviceCheckIsNotSupported() async throws {
        let expectedError = _GACAppCheckErrorUtil.unsupportedAttestationProvider("DeviceCheckProvider")
        
        fakeBackoffWrapper.backoffExpectation = expectation(description: "Backoff")
        let errorHandlerExpectation = expectation(description: "Error handler")
        
        fakeBackoffWrapper.defaultErrorHandler = { error in
            let nsError = error as NSError
            let expNSError = expectedError as NSError
            XCTAssertEqual(nsError.domain, expNSError.domain)
            XCTAssertEqual(nsError.code, expNSError.code)
            errorHandlerExpectation.fulfill()
            return .type1Day
        }
        
        fakeTokenGenerator.supported = false
        
        do {
            _ = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<GACAppCheckToken, Error>) in
                provider.getToken { token, error in
                    if let error = error {
                        continuation.resume(throwing: error)
                    } else if let token = token {
                        continuation.resume(returning: token)
                    } else {
                        continuation.resume(throwing: NSError(domain: "TestError", code: -1, userInfo: nil))
                    }
                }
            }
            XCTFail("Expected error")
        } catch let error as NSError {
            let expNSError = expectedError as NSError
            XCTAssertEqual(error.domain, expNSError.domain)
            XCTAssertEqual(error.code, expNSError.code)
        }
        
        await fulfillment(of: [fakeBackoffWrapper.backoffExpectation!, errorHandlerExpectation], timeout: 0.5)
        
        XCTAssertNil(fakeAPIService.passedDeviceToken)
        XCTAssertFalse(fakeTokenGenerator.generateTokenCalled)
        
        let opError = fakeBackoffWrapper.operationError as? NSError
        XCTAssertEqual(opError?.domain, (expectedError as NSError).domain)
        XCTAssertEqual(opError?.code, (expectedError as NSError).code)
        XCTAssertNil(fakeBackoffWrapper.operationResult)
    }
    
    func testGetTokenWhenDeviceTokenFails() async throws {
        let deviceTokenError = NSError(domain: "GACDeviceCheckProviderTests", code: -1, userInfo: nil)
        
        fakeBackoffWrapper.backoffExpectation = expectation(description: "Backoff")
        let errorHandlerExpectation = expectation(description: "Error handler")
        
        fakeBackoffWrapper.defaultErrorHandler = { error in
            let nsError = error as NSError
            XCTAssertEqual(nsError.domain, deviceTokenError.domain)
            XCTAssertEqual(nsError.code, deviceTokenError.code)
            errorHandlerExpectation.fulfill()
            return .type1Day
        }
        
        fakeTokenGenerator.supported = true
        fakeTokenGenerator.errorToReturn = deviceTokenError
        
        do {
            _ = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<GACAppCheckToken, Error>) in
                provider.getToken { token, error in
                    if let error = error {
                        continuation.resume(throwing: error)
                    } else if let token = token {
                        continuation.resume(returning: token)
                    } else {
                        continuation.resume(throwing: NSError(domain: "TestError", code: -1, userInfo: nil))
                    }
                }
            }
            XCTFail("Expected error")
        } catch let error as NSError {
            XCTAssertEqual(error.domain, deviceTokenError.domain)
            XCTAssertEqual(error.code, deviceTokenError.code)
        }
        
        await fulfillment(of: [fakeBackoffWrapper.backoffExpectation!, errorHandlerExpectation], timeout: 0.5)
        
        XCTAssertNil(fakeAPIService.passedDeviceToken)
        XCTAssertTrue(fakeTokenGenerator.generateTokenCalled)
        
        let opError = fakeBackoffWrapper.operationError as? NSError
        XCTAssertEqual(opError?.domain, deviceTokenError.domain)
        XCTAssertEqual(opError?.code, deviceTokenError.code)
        XCTAssertNil(fakeBackoffWrapper.operationResult)
    }
    
    func testGetTokenWhenAPIServiceFails() async throws {
        let apiError = NSError(domain: "GACDeviceCheckProviderTests", code: -1, userInfo: nil)
        
        fakeBackoffWrapper.backoffExpectation = expectation(description: "Backoff")
        let errorHandlerExpectation = expectation(description: "Error handler")
        
        fakeBackoffWrapper.defaultErrorHandler = { error in
            let nsError = error as NSError
            XCTAssertEqual(nsError.domain, apiError.domain)
            XCTAssertEqual(nsError.code, apiError.code)
            errorHandlerExpectation.fulfill()
            return .type1Day
        }
        
        fakeTokenGenerator.supported = true
        let deviceToken = Data()
        fakeTokenGenerator.tokenToReturn = deviceToken
        
        fakeAPIService.appCheckTokenResult = .failure(apiError)
        
        do {
            _ = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<GACAppCheckToken, Error>) in
                provider.getToken { token, error in
                    if let error = error {
                        continuation.resume(throwing: error)
                    } else if let token = token {
                        continuation.resume(returning: token)
                    } else {
                        continuation.resume(throwing: NSError(domain: "TestError", code: -1, userInfo: nil))
                    }
                }
            }
            XCTFail("Expected error")
        } catch let error as NSError {
            XCTAssertEqual(error.domain, apiError.domain)
            XCTAssertEqual(error.code, apiError.code)
        }
        
        await fulfillment(of: [fakeBackoffWrapper.backoffExpectation!, errorHandlerExpectation], timeout: 0.5)
        
        XCTAssertEqual(fakeAPIService.passedDeviceToken, deviceToken)
        XCTAssertEqual(fakeAPIService.passedLimitedUse, false)
        XCTAssertTrue(fakeTokenGenerator.generateTokenCalled)
        
        let opError = fakeBackoffWrapper.operationError as? NSError
        XCTAssertEqual(opError?.domain, apiError.domain)
        XCTAssertEqual(opError?.code, apiError.code)
        XCTAssertNil(fakeBackoffWrapper.operationResult)
    }
    
    func testGetLimitedUseTokenSuccess() async throws {
        fakeTokenGenerator.supported = true
        let deviceToken = Data()
        fakeTokenGenerator.tokenToReturn = deviceToken
        
        let validToken = GACAppCheckToken(token: "valid_token", expirationDate: Date.distantFuture, receivedAt: Date())
        fakeAPIService.appCheckTokenResult = .success(validToken)
        
        fakeBackoffWrapper.backoffExpectation = expectation(description: "Backoff")
        
        let token = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<GACAppCheckToken, Error>) in
            provider.getLimitedUseToken { token, error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else if let token = token {
                    continuation.resume(returning: token)
                } else {
                    continuation.resume(throwing: NSError(domain: "TestError", code: -1, userInfo: nil))
                }
            }
        }
        
        await fulfillment(of: [fakeBackoffWrapper.backoffExpectation!], timeout: 0.5)
        
        XCTAssertEqual(token.token, validToken.token)
        XCTAssertEqual(token.expirationDate, validToken.expirationDate)
        XCTAssertEqual(token.receivedAtDate, validToken.receivedAtDate)
        
        XCTAssertNil(fakeBackoffWrapper.operationError)
        let wrapperResult = fakeBackoffWrapper.operationResult as? GACAppCheckToken
        XCTAssertEqual(wrapperResult?.token, validToken.token)
        
        XCTAssertEqual(fakeAPIService.passedDeviceToken, deviceToken)
        XCTAssertEqual(fakeAPIService.passedLimitedUse, true)
        XCTAssertTrue(fakeTokenGenerator.generateTokenCalled)
    }
    
    // MARK: - Backoff tests
    
    func testGetTokenBackoff() async throws {
        fakeBackoffWrapper.isNextOperationAllowed = false
        fakeBackoffWrapper.backoffExpectation = expectation(description: "Backoff")
        
        do {
            _ = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<GACAppCheckToken, Error>) in
                provider.getToken { token, error in
                    if let error = error {
                        continuation.resume(throwing: error)
                    } else if let token = token {
                        continuation.resume(returning: token)
                    } else {
                        continuation.resume(throwing: NSError(domain: "TestError", code: -1, userInfo: nil))
                    }
                }
            }
            XCTFail("Expected error")
        } catch let error as NSError {
            let backoffError = fakeBackoffWrapper.backoffError as NSError
            XCTAssertEqual(error.domain, backoffError.domain)
            XCTAssertEqual(error.code, backoffError.code)
        }
        
        await fulfillment(of: [fakeBackoffWrapper.backoffExpectation!], timeout: 0.5)
        
        XCTAssertNil(fakeAPIService.passedDeviceToken)
        XCTAssertFalse(fakeTokenGenerator.generateTokenCalled)
    }
}
